### PR TITLE
arithmetic: Add Ctz function

### PR DESCRIPTION
**Pull request by eminemix**
Signed-off-by: Valentin Ilie <valentin.ilie@intel.com>

### CODE REVIEW

**Comment by eminemix** [comment]
First OCaml code ever written so this needs proper review. I've tested the functionality but I am not sure if this is actually a good patch.

File: ml-proto/src/arithmetic.ml [hunk 1]
```diff
@@ -50,7 +52,14 @@ struct
       | Abs -> Int.abs
       | Not -> Int.lognot
       | Clz -> fun i -> i  (* TODO *)
-      | Ctz -> fun i -> i  (* TODO *)
+      | Ctz -> fun i ->
+        if Int.to_int i == 0
```
Comment (rossberg-chromium) [inline ml-proto/src/arithmetic.ml]: `==` is physical equality in OCaml, you want `=`. Also, could say `i = Int.zero` instead, which is shorter.

File: ml-proto/src/arithmetic.ml [hunk 2]
```diff
@@ -50,7 +52,14 @@ struct
...
+        then Int.zero
+        else let rec
+          f j =
```
Comment (rossberg-chromium) [inline ml-proto/src/arithmetic.ml]: Weird formatting, put on previous line. Also, maybe rename `f` to `loop`.

File: ml-proto/src/arithmetic.ml [hunk 3]
```diff
@@ -50,7 +52,14 @@ struct
...
+          if (Int.to_int i) land j == 0
```
Comment (rossberg-chromium) [inline ml-proto/src/arithmetic.ml]: `=` not `==`

File: ml-proto/src/arithmetic.ml [hunk 4]
```diff
@@ -50,7 +52,14 @@ struct
...
+          then (Int.add Int.one (f (j lsl 1)))
+          else Int.zero in f 1
```
Comment (rossberg-chromium) [inline ml-proto/src/arithmetic.ml]: Weird formatting, put `in ...` on new line.

File: ml-proto/src/arithmetic.ml [hunk 5]
```diff
@@ -50,7 +52,14 @@ struct
...
       | Not -> Int.lognot
       | Clz -> fun i -> i  (* TODO *)
-      | Ctz -> fun i -> i  (* TODO *)
+      | Ctz -> fun i ->
+        if Int.to_int i == 0
+        then Int.zero
+        else let rec
+          f j =
+          if (Int.to_int i) land j == 0
+          then (Int.add Int.one (f (j lsl 1)))
```
Comment (rossberg-chromium) [inline ml-proto/src/arithmetic.ml]: Redundant outer parens.

**Comment by sunfishcode** [comment]
This is cool. Is there a way we can write a test for ctz while adding this code?

**Comment by lukewagner** [comment]
@sunfishcode I was hoping (myself soon or others if they go first) to add an "assert" `command` so that scripts could invoke a method and assert an expected result.  That should make it easy to throw up unit tests for ops like this.

**Comment by jfbastien** [comment]
@eminemix can you please join the [WebAssembly Community Group](https://www.w3.org/community/webassembly)? I can't find your name in the list. This is [required for significant contributions](https://github.com/WebAssembly/design/blob/master/Contributing.md) to WebAssembly.

**Comment by eminemix** [comment]
Sent v2.

@sunfishcode, @lukewagner 
I was thinking adding test code after implementing the clz.

@jfbastien 
I wall do that asap.

**Comment by rossberg-chromium** [comment]
Thanks for the patch. Please let me know when you have joined the Community Group, and I will merge your PR.

**Comment by eminemix** [comment]
This PR cannot be merged due to the issue pointed out by jfbastien.

This implementation may appear as part of other PRs where the author complies with the repo guidelines.

**Pull request closed without merging**